pilot/pkg/serviceregistry/kube/controller/ambient/statusqueue/queue.go

```diff
@@ -0,0 +1,132 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statusqueue
+
+import (
+	"k8s.io/apimachinery/pkg/types"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/krt"
+	"istio.io/istio/pkg/log"
+)
+
+type StatusQueue struct {
+	queue controllers.Queue
+
+	// reporters is a mapping of unique controller name -> status information.
+	// Note: this is user facing in the fieldManager!
+	reporters map[string]statusReporter
+}
+
+// statusItem represents the objects stored on the queue
+type statusItem struct {
+	Key      string
+	Reporter string
+}
+
+// NewQueue builds a new status queue.
+func NewQueue() *StatusQueue {
+	sq := &StatusQueue{
+		reporters: make(map[string]statusReporter),
+	}
+	sq.queue = controllers.NewQueue("ambient status",
+		controllers.WithGenericReconciler(sq.reconcile),
+		controllers.WithMaxAttempts(5))
+	return sq
+}
+
+// Run starts the queue, which will process items until the channel is closed
+func (q *StatusQueue) Run(stop <-chan struct{}) {
+	q.queue.Run(stop)
+}
+
+// reconcile processes a single queue item
+func (q *StatusQueue) reconcile(raw any) error {
+	key := raw.(statusItem)
+	log := log.WithLabels("key", key.Key)
+	log.Debugf("reconciling status")
+
+	reporter, f := q.reporters[key.Reporter]
+	if !f {
+		log.Fatalf("impossible; an item was enqueued with an unknown reporter")
+	}
+	obj, f := reporter.getObject(key.Key)
+	if !f {
+		log.Infof("object is removed, no action needed")
+		return nil
+	}
+	// Fetch the client to apply patches, and the set of current conditions
+	patcher, currentConditions := reporter.patcher(obj)
+	// Turn the conditions into a patch. Using currentConditions, this will delete and conditions
+	// that are owned by us, but not currently set (while not deleting conditions owned by other controllers).
+	// An alternative approach would be to use SSA. The problem with this is it requires storing ManagedFields; Istio uses an
+	// optimization to drop these from objects, which has a decent savings in memory usage of Istiod.
+	//
+	// This gives us essentially the same behavior, with the caveat of one race condition:
+	// the object can change between when we fetch the currentConditions and apply the patch.
+	// * Condition was there, but is now removed: we will issue a `$patch: delete` on it, which does nothing. No problem.
+	// * Condition was not there, but now it was added: clearly some other controller is writing the same type as us, which is not really allowed.
```

Comment:
What happens when we run revisioned istiod with Ambient and this status writing? Should there be a leader election which is cluster scoped to prevent racy updates from different istiods?

Reply:
Ah yeah for sure - thanks for the callout. Forgot to add that!